### PR TITLE
Add wrapping to long diff lines to fix #1827

### LINKED ISSUE #1827: Wrapping long lines in code views
Currently, Gitea does not wrap long lines of code, making it harder to look at code and diffs that contains lines that extend beyond the screen width.

Is there a general interest in having long lines automatically wrapped? It's a simple CSS tweak (similar to what I've done [here](https://github.com/StylishThemes/GitHub-code-wrap/blob/master/github-code-wrap.css) for GitHub):

Before:

After:

This might not work too well on small screens, where I think it might be better to not wrap those lines (horizontal scrolling is easier on touch).

### PR DESCRIPTION
**Pull request by mjwwit**
This PR adds line wrapping to the unified diff view (for long lines) (#1827). Wrapping was already done in the split view.
An additional change has been made to align the line-numbers of wrapped diff lines to the top of the line.
gitea-long-diff-line-wrapped

EDIT: Closes #1827

### CODE REVIEW

**Comment by bkcsoft** [comment]
Maybe have a checkbox toggle? For those that don't want wrapping :)

~Also Closes #1827~

**Comment by mjwwit** [comment]
The wrapping was already default behavior for the split view, are you sure you want a setting for this?
If you do want a setting, where should it be located? (we don't really have a user preferences view)
Also, should it store the preference server-side or simply keep it client-side?

File: public/css/index.css [hunk 1]
```diff
@@ -2063,6 +2063,7 @@ footer .ui.language .menu {
   color: #A7A7A7;
   background: #fafafa;
   width: 1%;
+  vertical-align: top;
```
Comment (ethantkoenig) [inline public/css/index.css]: Please don't edit `index.css` directly, instead edit the appropriate `.less` file

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

**Comment by silverwind** [comment]
> If you do want a setting, where should it be located?

Maybe beside 'View File' in the header?

BTW I also noticed an error in the `.wrap` CSS: The current `break-word` value for `word-break` is invalid. It should be `break-all`. I recommend these rules for it:

```
white-space: pre-wrap;
word-break: break-all;
overflow-wrap: break-word;
```

Oh, and I also think this change should apply to regular code view, not just diffs.

File: public/css/index.css (hunk 1 above)
Comment (mjwwit) [reply public/css/index.css]: Ugh... I hate generated code in VCS... I'll fix it though, good spot!

**Comment by mjwwit** [comment]
I assumed the `wrap` class would have a proper implementation, as it turns out I was wrong. I have to keep the non-standard `word-break: break-word;` as well though, as it's needed for WebKit to function properly.

I'm not sure I agree with applying this to normal code views as well. And if we decide to do this, I'd do so using a user preference.

**Comment by silverwind** [comment]
Looks like `word-break: break-word` [is superceded](https://stackoverflow.com/a/15137272/808699) by `overflow-wrap: break-word`, so these two should be equivalent. I can at least say that the former isn't recognized by Firefox any more:



Thought, from the linked description, it sounds that `break-all` is what we want, and `overflow-wrap` is unnecessary.

**Comment by mjwwit** [comment]
Interesting, reading the W3C spec leads me to believe `overflow-wrap` can be used in in addition to `word-break`. I don't think one is superseded by the other.

`word-break: break-word;` should only be used (and will only be accepted) by WebKit. Other browsers correctly implemented the standard `word-break: break-all;` value.

**Comment by silverwind** [comment]
Both are valid in WebKit:



The [spec](https://www.w3.org/TR/css-text-3) seems pretty clear to me. `word-wrap` does not list it as a valid property value and `overflow-wrap` contains this note:

> For legacy reasons, UAs must treat ‘word-wrap’ as an alternate name for the ‘overflow-wrap’ property, as if it were a shorthand of ‘overflow-wrap’.

File: public/less/_base.less [hunk 2]
```diff
@@ -26,7 +26,15 @@ pre, code {
...
 //		white-space: -normal;      /* Opera 4-6 */
 //		white-space: -o-normal;    /* Opera 7 */
+
+		/* This is the dangerous one in WebKit, as it breaks things wherever */
+		word-break: break-all;
+		/* Instead use this non-standard one: */
 		word-break: break-word;
+
+		/* These are technically the same, but use both */
+  	overflow-wrap: break-word;
```
Comment (lafriks) [inline public/less/_base.less]: Something seems to be wrong with tabs

**Review by lafriks — CHANGES_REQUESTED** [review]

**Review by ethantkoenig — APPROVED** [review]

**Comment by mjwwit** [comment]
> The spec seems pretty clear to me. `word-wrap` does not list it as a valid property value and `overflow-wrap`...

But we're talking about `word-break` and `overflow-wrap`, not `word-wrap`.

**Comment by lunny** [comment]
need @lafriks approval.

**Comment by silverwind** [comment]
> But we're talking about word-break and overflow-wrap, not word-wrap.

Yeah, my error, I meant `word-break`. Anyways, I think you should try my suggestion in https://github.com/go-gitea/gitea/pull/1954#issuecomment-307868910, it should be all that's needed to achieve consistent cross-browser wrapping.

**Review by lafriks — APPROVED** [review]

File: public/less/_repository.less [hunk 3]
```diff
@@ -203,15 +203,15 @@
 				padding-top: 8px;
 				padding-bottom: 8px;
 			}
-            td.message .isSigned {
-                cursor: default;
-            }
+						td.message .isSigned {
```
Comment (lafriks) [inline public/less/_repository.less]: Too much tabs

File: public/less/_repository.less [hunk 4]
```diff
@@ -203,15 +203,15 @@
...
+								cursor: default;
+						}
 			tr:hover {
 				background-color: #ffffEE;
 			}
-            .jumpable-path {
-                color: #888;
-            }
+						.jumpable-path {
```
Comment (lafriks) [inline public/less/_repository.less]: Too much tabs

File: public/less/_repository.less [hunk 5]
```diff
@@ -678,7 +678,7 @@
 			}
 			textarea {
 				height: 200px;
-                font-family: "Consolas", monospace;
+								font-family: "Consolas", monospace;
```
Comment (lafriks) [inline public/less/_repository.less]: Too much tabs

File: public/less/_repository.less [hunk 6]
```diff
@@ -810,34 +810,34 @@
 				width: 140px;
 			}
 		}
-        td.sha .sha.label {
-            margin: 0;
-        }
+				td.sha .sha.label {
```
Comment (lafriks) [inline public/less/_repository.less]: Too much tabs

File: public/less/_repository.less [hunk 7]
```diff
@@ -810,34 +810,34 @@
...
-        }
-        &.isSigned.isVerified{
-            border: 1px solid #21BA45;
-            background: #21BA4518;
-            .detail.icon{
-                border-left: 1px solid #21BA4580;
-            }
-        }
-    }
+		#commits-table td.sha .sha.label, #repo-files-table .sha.label{
```
Comment (lafriks) [inline public/less/_repository.less]: Too much tabs in all levels

File: public/less/_repository.less [hunk 8]
```diff
@@ -1237,16 +1238,16 @@
...
-		    border-left: 1px solid #A3C293;
-		    border-right: 1px solid #A3C293;
-	    }
-	    &.top:not(.positive){
-		    border-top: 1px solid #A3C293;
-	    }
-        &:not(.positive):last-child {
-            border-bottom: 1px solid #A3C293;
-        }
+				&:not(.positive){
```
Comment (lafriks) [inline public/less/_repository.less]: Too much tabs

File: public/less/_repository.less [hunk 9]
```diff
@@ -1304,7 +1305,7 @@
 }
 
 .issue-actions {
-    display: none;
+		display: none;
```
Comment (lafriks) [inline public/less/_repository.less]: One tab too much :)

**Review by lafriks — CHANGES_REQUESTED** [review]
When tabs are fixed than LG-TM

**Comment by mjwwit** [comment]
This whole formatting problem is caused by the .editorconfig file stating that indentation should be done using 4 spaces, but the less files are using tabs. Which should it be?

**Comment by mjwwit** [comment]
Lack of response leads me to believe nobody really cares if indentation is done using spaces or tabs, so I'll go with 4 spaces.

**Comment by silverwind** [comment]
Getting back to the topic of having it also apply on regular code views: Why should it only wrap in diffs? If you have a long line in a diff, it'll be a long line in the regular code view. I'd rather have both wrap for consistence.

Of course, a checkbox toggle in the header would be ideal, but that can come later, imho.

**Comment by mjwwit** [comment]
This may just be me, but I generally hate wrapping in code. The only time I do find it useful is when I need to see the difference between two pieces of code in a single glance. The wrapping is especially useful in a side-by-side diff, where the horizontal space is extra limited. I'm not against implementing wrapping in normal code-views as well, but I'd rather build the toggle setting with it so I can turn it off. Would you be willing to wait for the normal code view wrapping until that time?

**Comment by lafriks** [comment]
@mjwwit in code view and diff view code wrapping is needed as otherwise if code block is larger than screen you have to scroll to the bottom of code block to be able to scroll to right

**Comment by silverwind** [comment]
I'm fine if you want to do it later with the toggle. Having it in diffs is already an improvement. Side-by-side diffs already wrap without this patch for me, by the way.

Also agree with @lafriks, horizontal scroll on big files is generally a bad UX, but there are options like using the arrow keys to scroll.

**Review by silverwind — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
I vote for 4 spaces for tabbing just wanted more opinion from other Gitea developers. If other are also OK with that than it is LG-TM

**Comment by bkcsoft** [comment]
@lafriks LG-TM on 4 space. Haven't reviewed this one though so no real LG-TM from me ;)

**Comment by lunny** [comment]
hi, please rebase.

**Comment by lafriks** [comment]
Rebased and tested, LGTM

**Pull request closed without merging**